[DOC] Tempo 2.3 release notes (#3089)
grafana/tempo: docs/sources/tempo/release-notes/v2-3.md

---
title: Version 2.3 release notes
menuTitle: V2.3
description: Release notes for Grafana Tempo 2.3
weight: 45
---

# Version 2.3 release notes

The Tempo team is pleased to announce the release of Tempo 2.3.

This release gives you:

* New [structural operators]({{< relref "../traceql#experimental-structural" >}}) in TraceQL:
    * Ancestor (`<<`)
    * Parent (`<`)

>     * Not-child (`!>`) (experimental)

Reviewer: not ancestor `!<<` and not descendant have also been added `!>>`

>     * not parent (`!<`) (experimental)
>     * not siblings (`!~`) (experimental)

Author: ```suggestion
    * Not ancestor (`!<<`) (experimental)
    * Not parent (`!<`) (experimental)
    * Not siblings (`!~`) (experimental)
```

* TraceQL support for [searching quoted attributes]({{< relref "../traceql#experimental-structural" >}}) (for example, `{ span."name with spaces" = "foo" }`).
* [Dedicated attribute columns]({{< relref "../operations/dedicated_columns" >}}) for improving query performance on your most frequently queried attributes

Tempo 2.3 makes [vParquet3]({{< relref "../configuration/parquet" >}}), a Parquet version designed to be more compatible with other Parquet implementations, available as a production-ready option.

> This block format is required for using structural operators and improves query performance relative to previous formats.

Reviewer: 2.2 does structural operators but 2.3 is faster

Author: ```suggestion
This block format improves query performance relative to previous formats. Structural operators require vParquet3.
```

Reviewer: all structural operators (including the new ones) work on both vp2 and vp3. there are, however, optimizations in 2.3 which make them a bit faster (with more to come).

Read the [**Tempo 2.3 blog post**](NEED-LINK) for more examples and details about these improvements.

These release notes highlight the most important features and bugfixes. For a complete list, refer to the [Tempo changelog](https://github.com/grafana/tempo/releases).

## Features and enhancements

The most important features and enhancements in Tempo 2.3 are highlighted below.

### vParquet3 promoted to stable

[vParquet3]({{< relref "../configuration/parquet" >}}), an updated Parquet block format introduced in Tempo 2.2, has been promoted to stable and is production ready.

> We're excited about vParquet3 relative to prior formats because of its support for dedicated attribute columns, which help speed up queries on your most queried attributes. [[PR 2649](https://github.com/grafana/tempo/pull/2649)]

Reviewer: we've actually found that the best is a mix of your largest attributes and your most queried

To read more about the design of vParquet3, refer to [the design proposal](https://github.com/grafana/tempo/blob/main/docs/design-proposals/2023-05%20vParquet3.md).
For general information, refer to [the Apache Parquet schema]({{< relref "../operations/schema" >}}).

#### Dedicated attribute columns

> Dedicated attribute columns improve query performance by storing the most frequently used attributes in their own columns, rather than in the generic attribute key-value list.

Reviewer: same comment as above

Author: ```suggestion
Dedicated attribute columns improve query performance by storing the largest and most frequently used attributes in their own columns, rather than in the generic attribute key-value list.
```

Introduced with vParquet3, dedicated attribute columns are only available when using this storage format.

To learn more about this new feature, refer to the [Dedicated attribute columns]({{< relref "../operations/dedicated_columns" >}}) documentation.

### TraceQL

Unique to Tempo, TraceQL is a query language that lets you perform custom queries into your tracing data.
To learn more about the TraceQL syntax, see the [TraceQL documentation]({{< relref "../traceql" >}}).
For information on planned future extensions to the TraceQL language, refer to [future work]({{< relref "../traceql/architecture" >}}).

We’ve made the following improvements to TraceQL:

* Added two structural operators, ancestor (`<<`) and parent (`<`) ([documentation]({{< relref "../traceql#experimental-structural" >}})) [[PR 2877](https://github.com/grafana/tempo/pull/2877)]

> * Added experimental not structural operators: `!>`, `!<`, and `!~` ([documentation]({{< relref "../traceql#experimental-structural" >}})) [[PR 2993](https://github.com/grafana/tempo/pull/2993)]

Reviewer: same comment as above

Author: ```suggestion
* Added experimental not structural operators: `!>>`, `!>`, `!<<`, `!<`, and `!~` ([documentation]({{< relref "../traceql#experimental-structural" >}})) [[PR 2993](https://github.com/grafana/tempo/pull/2993)]
```

> * Added quoted attribute syntax so you can search for attributes with terminal characters like periods (`.`) in their names ([documentation]({{< relref "../traceql#experimental-structural" >}})) [[PR 3004](https://github.com/grafana/tempo/pull/3004)]

Reviewer: periods were always fine. i think space ` ` is the most obvious character we've enabled the use of

Author: ```suggestion
* Added quoted attribute syntax so you can search for attributes with terminal characters like spaces in their names ([documentation]({{< relref "../traceql#experimental-structural" >}})) [[PR 3004](https://github.com/grafana/tempo/pull/3004)]
```

Reviewer: here is the list of characters we previously terminated an attribute name on in traceql. if these characters showed up in an attribute name they would have correctly returned when retrieving the trace by id but they were not referenceable in traceql:

[whitespace chars](https://github.com/grafana/tempo/blob/main/pkg/traceql/lexer.go#L345)
[other chars](https://github.com/grafana/tempo/blob/main/pkg/traceql/lexer.go#L350)

maybe we just say "space and other special chars?"

* Improved the performance of TraceQL [`select()` queries]({{< relref "../traceql#selection" >}}). Metrics-summary now also correctly handles missing attributes. [[PR 2765](https://github.com/grafana/tempo/pull/2765)]

> * Added support for searching by OpenTelemetry’s [span status message ](https://github.com/open-telemetry/opentelemetry-proto/blob/afcd2aa7f728216d5891ffc0d83f09f0278a6611/opentelemetry/proto/trace/v1/trace.proto#L260)using `statusMessage` intrinsic attribute (read [documentation]({{< relref "../traceql#intrinsic-fields" >}})) [[PR 2848](https://github.com/grafana/tempo/pull/2848)]

Author: ```suggestion
* Added support for searching by OpenTelemetry’s [span status message ](https://github.com/open-telemetry/opentelemetry-proto/blob/afcd2aa7f728216d5891ffc0d83f09f0278a6611/opentelemetry/proto/trace/v1/trace.proto#L260)using `statusMessage` intrinsic attribute ([documentation]({{< relref "../traceql#intrinsic-fields" >}})) [[PR 2848](https://github.com/grafana/tempo/pull/2848)]
```

---
title: Version 2.3 release notes
menuTitle: V2.3
description: Release notes for Grafana Tempo 2.3
weight: 45
---

# Version 2.3 release notes

The Tempo team is pleased to announce the release of Tempo 2.3.

This release gives you:

* New [structural operators]({{< relref "../traceql#experimental-structural" >}}) in TraceQL:
    * Ancestor (`<<`)
    * Parent (`<`)
    * Not-child (`!>`) (experimental)
    * not parent (`!<`) (experimental)
    * not siblings (`!~`) (experimental)

* TraceQL support for [searching quoted attributes]({{< relref "../traceql#experimental-structural" >}}) (for example, `{ span."name with spaces" = "foo" }`).
* [Dedicated attribute columns]({{< relref "../operations/dedicated_columns" >}}) for improving query performance on your most frequently queried attributes

Tempo 2.3 makes [vParquet3]({{< relref "../configuration/parquet" >}}), a Parquet version designed to be more compatible with other Parquet implementations, available as a production-ready option.
This block format is required for using structural operators and improves query performance relative to previous formats.

> Read the [**Tempo 2.3 blog post**](NEED-LINK) for more examples and details about these improvements.

Author: Doc-validator won't pass until the blog post link is here.

Reviewer: we're in a bit of a chicken/egg situation here. normally i link to the release notes. i was hoping they'd be in place today/tomorrow morning so i could add the link to the blogpost

Author: We can always publish these and update the link to the blog post. We know what the release notes link will be -- it's on the front of the PR. 
```suggestion
Read the **Tempo 2.3 blog post** for more examples and details about these improvements.
```

---
title: Version 2.3 release notes
menuTitle: V2.3
description: Release notes for Grafana Tempo 2.3
weight: 45
---

# Version 2.3 release notes

The Tempo team is pleased to announce the release of Tempo 2.3.

This release gives you:

* New [structural operators]({{< relref "../traceql#experimental-structural" >}}) in TraceQL:
    * Ancestor (`<<`)
    * Parent (`<`)
    * Not-child (`!>`) (experimental)
    * Not descendant (`!>>`) (experimental)
    * Not ancestor (`!<<`) (experimental)
    * Not parent (`!<`) (experimental)
    * Not siblings (`!~`) (experimental)

* TraceQL support for [searching quoted attributes]({{< relref "../traceql#experimental-structural" >}}) (for example, `{ span."name with spaces" = "foo" }`).
* [Dedicated attribute columns]({{< relref "../operations/dedicated_columns" >}}) for improving query performance on your most frequently queried attributes

> Tempo 2.3 makes [vParquet3]({{< relref "../configuration/parquet" >}}), a Parquet version designed to be more compatible with other Parquet implementations, available as a production-ready option.

Reviewer: ```suggestion
Tempo 2.3 introduces [vParquet3]({{< relref "../configuration/parquet" >}}), a Parquet version designed to be more compatible with other Parquet implementations, available as a production-ready option.
```

* Added `TempoUserConfigurableOverridesReloadFailing` alert [[PR 2784](https://github.com/grafana/tempo/pull/2784)]
* Added a metrics-generator configuration option to enable/disable X-Scope-OrgID headers on remote write to better support single-tenant setups. [[PR 2974](https://github.com/grafana/tempo/pull/2974)]
* Allowed metrics-generator ingestion slack to be configured on a per-tenant, rather than global, basis [[PR 2589](https://github.com/grafana/tempo/pull/2589)]
* Added several metrics-generator fields to user-configurable overrides, including histogram buckets, collection-interval, and span metrics filter policies [[PR 2906](https://github.com/grafana/tempo/pull/2906), [2711](https://github.com/grafana/tempo/pull/2711), [2928](https://github.com/grafana/tempo/pull/2928), [2899](https://github.com/grafana/tempo/pull/2899)]

## Upgrade considerations

When [upgrading]({{< relref "../setup/upgrade" >}}) to Tempo 2.3, be aware of these considerations and breaking changes.

### Transition to vParquet 3

> Although the vParquet3 format isn't yet the default, we believe it's production ready and highly recommend switching to it for improved query performance and [dedicated attribute columns]({{< relref "../operations/dedicated_columns" >}}).

Reviewer: ```suggestion
Although the vParquet3 format isn't yet the default, it's production ready and we highly recommend switching to it for improved query performance and [dedicated attribute columns]({{< relref "../operations/dedicated_columns" >}}).
```
